engine/verification/finder/engine.go

```diff
@@ -314,64 +341,199 @@ func (e *Engine) checkCachedReceipts() {
+	receiptID := receiptDataPack.Receipt.ID()
+	resultID := receiptDataPack.Receipt.ExecutionResult.ID()
+	blockID := receiptDataPack.Receipt.ExecutionResult.BlockID
+	log := e.log.With().
+		Hex("origin_id", logging.ID(receiptDataPack.OriginID)).
+		Hex("block_id", logging.ID(blockID)).
+		Hex("receipt_id", logging.ID(receiptID)).
+		Hex("result_id", logging.ID(resultID)).Logger()
+
+	ok := e.pendingReceipts.Add(receiptDataPack)
+	if !ok {
+		log.Debug().Msg("drops adding duplicate receipt to pending mempool")
+		return
+	}
+
+	// marks receipt pending for its block ID
+	err := e.pendingReceiptIDsByBlock.Append(blockID, receiptID)
+	if err != nil {
+		log.Debug().Err(err).Msg("could not append receipt to receipt-ids-by-block mempool")
+	}
+
+	// records the execution receipt id based on its result id
+	err = e.receiptIDsByResult.Append(resultID, receiptID)
+	if err != nil {
+		log.Debug().Err(err).Msg("could not add receipt id to receipt-ids-by-result mempool")
+		return
+	}
+
+	log.Info().Msg("cached execution receipt added to pending mempool")
+}
+
+// pendingToReady receives a list of receipt identifiers and moves all their corresponding receipts
+// from pending to ready mempools.
+// blockID is the block identifier that all receipts are pointing to.
+func (e *Engine) pendingToReady(receiptIDs flow.IdentifierList, blockID flow.Identifier) {
+	for _, receiptID := range receiptIDs {
+		// retrieves receipt from pending mempool
+		rdp, ok := e.pendingReceipts.Get(receiptID)
+		log := e.log.With().
+			Hex("block_id", logging.ID(blockID)).
+			Hex("receipt_id", logging.ID(receiptID)).
+			Logger()
+		if !ok {
+			log.Debug().Msg("could not retrieve receipt from pending receipts mempool")
+			continue
+		}
+
+		resultID := rdp.Receipt.ExecutionResult.ID()
+		log = log.With().
+			Hex("result_id", logging.ID(resultID)).
+			Logger()
+
+		// NOTE: this anonymous function is solely for sake of encapsulating a block of code
+		// for opentracing. To avoid closure, it should NOT encompass any goroutine involving rdp.
+		func() {
+			var span opentracing.Span
+			span, _ = e.tracer.StartSpanFromContext(rdp.Ctx, trace.VERFindCheckPendingReceipts)
+			defer span.Finish()
+
+			// moves receipt from pending to ready mempool
+			removed := e.pendingReceipts.Rem(receiptID)
+			log.Debug().
+				Bool("removed", removed).
+				Msg("removes receipt from pending receipts")
+
+			added := e.readyReceipts.Add(rdp)
+			log.Debug().
+				Bool("added", added).
+				Msg("adds receipt to ready receipts")
+		}()
+	}
+}
+
+// discardPendingReceipts receives a list of receipt ids, and removes
+// all receipts from the pending receipts mempool and marks their execution result as discarded.
+// blockID is the block identifier that all receipts are pointing to.
+//
+// finder engine discards a receipt if it is not staked at block id of that receipt.
+func (e *Engine) discardPendingReceipts(receiptIDs flow.IdentifierList, blockID flow.Identifier) {
```

Comment:
The receipts are no longer pending, since we've received the block. We could rename it to `discardReceiptsFromPending` instead.

Reply:
done in https://github.com/onflow/flow-go/pull/245/commits/d8e14cdc4dea249478df33599b506ec2aa2beab7